src/main/java/org/openlmis/buq/service/buq/BottomUpQuantificationService.java

```diff
@@ -0,0 +1,232 @@
+import static org.openlmis.buq.i18n.MessageKeys.ERROR_PROCESSING_PERIOD_NOT_FOUND;
+import static org.openlmis.buq.i18n.MessageKeys.ERROR_PROGRAM_NOT_FOUND;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.openlmis.buq.domain.buq.BottomUpQuantification;
+import org.openlmis.buq.domain.buq.BottomUpQuantificationLineItem;
+import org.openlmis.buq.domain.buq.BottomUpQuantificationStatus;
+import org.openlmis.buq.domain.buq.BottomUpQuantificationStatusChange;
+import org.openlmis.buq.dto.buq.BottomUpQuantificationDto;
+import org.openlmis.buq.dto.referencedata.ApprovedProductDto;
+import org.openlmis.buq.dto.referencedata.FacilityDto;
+import org.openlmis.buq.dto.referencedata.ProcessingPeriodDto;
+import org.openlmis.buq.dto.referencedata.ProgramDto;
+import org.openlmis.buq.exception.ContentNotFoundMessageException;
+import org.openlmis.buq.exception.ValidationMessageException;
+import org.openlmis.buq.i18n.MessageKeys;
+import org.openlmis.buq.repository.buq.BottomUpQuantificationRepository;
+import org.openlmis.buq.service.referencedata.ApprovedProductReferenceDataService;
+import org.openlmis.buq.service.referencedata.FacilityReferenceDataService;
+import org.openlmis.buq.service.referencedata.PeriodReferenceDataService;
+import org.openlmis.buq.service.referencedata.ProgramReferenceDataService;
+import org.openlmis.buq.util.AuthenticationHelper;
+import org.openlmis.buq.util.FacilitySupportsProgramHelper;
+import org.openlmis.buq.util.Message;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@SuppressWarnings("PMD.TooManyMethods")
+@Service
+public class BottomUpQuantificationService {
+
+  @Autowired
+  private AuthenticationHelper authenticationHelper;
+
+  @Autowired
+  private FacilityReferenceDataService facilityReferenceDataService;
+
+  @Autowired
+  private FacilitySupportsProgramHelper facilitySupportsProgramHelper;
+
+  @Autowired
+  private ProgramReferenceDataService programReferenceDataService;
+
+  @Autowired
+  private PeriodReferenceDataService periodReferenceDataService;
+
+  @Autowired
+  private ApprovedProductReferenceDataService approvedProductReferenceDataService;
+
+  @Autowired
+  private BottomUpQuantificationRepository bottomUpQuantificationRepository;
+
+  /**
+   * Prepares given bottom-up quantification if possible.
+   *
+   * @param facilityId         Facility ID.
+   * @param programId          Program ID.
+   * @param processingPeriodId Processing Period ID.
+   * @return Prepared bottom-up quantification.
+   */
+  public BottomUpQuantification prepare(UUID facilityId, UUID programId,
+       UUID processingPeriodId) {
+    if (null == facilityId || null == programId || null == processingPeriodId) {
+      throw new ValidationMessageException(
```

Comment:
Generally, it is good idea to list what is missing. As I understand, this will just throw "exception: missingParameters" right now? You would need to check the source code to know what is missing, to start guessing what goes wrong in other parts of the app.

If this would be used only during development time, then I tend to use `java.util.Objects#requireNonNull(facilityId, "facilityId requored")`.
If this needs to print error for FE, then information about what exact parameter is missing would be useful I think.

Reply:
I changed it to send to FE a message with a list of which parameter is missing.